### PR TITLE
Support user supplied keypairs

### PR DESCRIPTION
**Pull request by fzgregor**
- renamed DynKeyPair to KeyPair - might not be dynamic anymore
- made KeyPair public added From implementation
- moved serialized key pair into KeyPair struct
- introduced Certificate::from_params_and_keypair

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -304,16 +303,20 @@ fn dt_to_generalized(dt :&DateTime<Utc>) -> GeneralizedTime {
 }
 
 impl Certificate {
-	/// Generates a new self-signed certificate from the given parameters
-	pub fn from_params(params :CertificateParams) -> Self {
-		let (key_pair, key_pair_serialized) = DynKeyPair::generate(&params.alg);
+	/// Generates a new certificate from the given parameters and key pair
+	pub fn from_params_and_keypair(params :CertificateParams, key_pair :KeyPair) -> Self {
```
Comment (est31) [inline src/lib.rs]: Instead of this function, could you add a `key_pair :Option<KeyPair>` field to `CertificateParams` instead?

File: src/lib.rs [hunk 2]
```diff
@@ -577,46 +583,65 @@ enum SignAlgo {
...
+	EcKp(EcdsaKeyPair, Vec<u8>),
+	/// A Ed25519 key pair
+	EdKp(Ed25519KeyPair, Vec<u8>),
+}
+
+impl From<&[u8]> for KeyPair {
+	fn from(pkcs8 :&[u8]) -> KeyPair {
+		use std::iter::FromIterator;
+
+		if let Ok(edkp) = Ed25519KeyPair::from_pkcs8(Input::from(pkcs8)) {
```
Comment (est31) [inline src/lib.rs]: Could you add a `let input = Input::from(pkcs8);` to the front and then use it in each function? This makes the lines shorter. Possibly you'll have to clone it, most likely you won't.

File: src/lib.rs (hunk 1 above)
Comment (fzgregor) [reply src/lib.rs]: Sure thing. Like this https://github.com/est31/rcgen/commit/33ea827ce1dcbee50d88dea2adfd8b4c548dbc2c#diff-b4aea3e418ccdb71239b96952d9cddb6L307?

File: src/lib.rs (hunk 2 above)
Comment (fzgregor) [reply src/lib.rs]: Did that. I also move the pkc8 vector cloning in front: https://github.com/est31/rcgen/commit/33ea827ce1dcbee50d88dea2adfd8b4c548dbc2c#diff-b4aea3e418ccdb71239b96952d9cddb6R591

File: src/lib.rs (hunk 1 above)
Comment (est31) [reply src/lib.rs]: Yup, that's good.

**Comment by fzgregor** [comment]
Please note that I also moved the `serialize_*` functions into `KeyPair` and made the `generate()` function public. As `KeyPair` is public now this feels more correct to me. I'm not sure how you think about this. Given your comment at https://github.com/est31/rcgen/issues/7#issuecomment-500831262 you might feel different about this.

**Comment by fzgregor** [comment]
Oh wow, you're fast.

**Comment by est31** [comment]
@fzgregor making the `generate` function public hurts nobody and the code is needed for certificate generation. In my comment you linked I was only worried that I didn't want to expand scope of this crate infinitely :). Or rather, that I'm not sure yet whether I do want to expand it :).

**Comment by fzgregor** [comment]
Glad to hear that.

**Pull request merged**